geps/gep-1911.md

```diff
@@ -0,0 +1,181 @@
+
+## API Semantics
+
+A Gateway implementation MUST recognize the Kubernetes Standard Application Protocols ([KEP-3726][kep-3726]) for specifying the protocol for a backend reference in a Gateway API `*Route` resource
+
+Thus when a `*Route` points to a Kubernetes Service resource the backend protocol for each port can be specified by
+
+- setting the `appProtocol` field on the Kubernetes `Service`
+- setting the `appProtocol` field on an `Endpoint`/`EndpointSlice` object associated with a Kubernetes `Service`
+
+At the moment there exists three defined constants:
+
+- `kubernetes.io/h2c` - HTTP/2 over cleartext as described in [RFC7540](https://www.rfc-editor.org/rfc/rfc7540)
+- `kubernetes.io/ws` - WebSocket over cleartext as described in [RFC6445](https://www.rfc-editor.org/rfc/rfc6455)
+- `kubernetes.io/wss` - WebSocket over TLS as described in [RFC6455](https://www.rfc-editor.org/rfc/rfc6455)
+
+### New Protocols & Reserved Prefix
+
+To add support for a new protocol it should first become a Kubernetes Standard Application Protocol by updating the [KEP-3726][kep-3726]. [KEP-3726][kep-3726] also states the `appProtocol` field accepts a domain-prefixed implementation specific value. Thus, if the suggested protocol is not suited to have a `kubernetes.io/*` prefix, then the Gateway API MAY support the new protocol using it's own prefix `gateway.networking.k8s.io/*`.  Please make a PR to this GEP.
+
+For example we may want to add a sentinel `appProtocol` value that prevents Gateway implementations from discovering the protocol of the application. Instead they should just refer to the Service's `protocol` field. Such a constant was rejected upstream (https://github.com/kubernetes/enhancements/pull/4106) but as an example it could be defined in a future addition to this GEP as `gateway.networking.k8s.io/no-sniff`.
+
+### Default Protocols
+
+If a Service `appProtocol` isn't specified an implementation MAY infer the backend protocol through its own means. Implementations MAY infer the protocol from the `Route` type referring to the backend Service.
+
+### Multiple Protocols on the same port
+
+Only the Kubernetes `Service` `protocol` field supports multiple protocols on the same port. See the details in [KEP-1435][kep-1435].
+
+Implementations MUST recognize support for `TCP` and `UDP` multiplexing on the same port.
+
+Currently Kubernetes `Service` API only allows different `appProtocol` values for the same port when `protocol` fields differs. At this time there seems to be interest in changing `appProtocol` to be a list in order to faciliate this use-case.
+
+### Unsupported/Incompatible Protocols
+
+If a Route attached to a Gateway is not able to send traffic to the backend using the specified protocol then the backend is considered invalid. Implementations MUST set `ResolvedRefs` condition to `False` with the Reason `UnsupportedProtocol`.
+
+Implementations MUST supported the following combinations below:
```

Comment:
I don't think that we can require implementations to support these combinations. We can recommend that they do, or even just state these combinations have well-defined interactions with Gateway API and should be supported wherever possible.

Reply:
Should I change the entire GEP to be `RECOMMENDED` ? line 37 has a `MUST`

Reply:
this section has some language that might help, talking about Listener protocols so a similar idea: https://github.com/kubernetes-sigs/gateway-api/blob/ea382e7a989d2849ce1a76d47a60bd5f5ea754e5/apis/v1beta1/gateway_types.go#L260-L266

(though the line referencing a table in the linked godoc might need to be removed now)

I think in other places `MAY` is used in similar contexts

If we anticipate protocol/kind combos in the tables to be tested in conformance then could add a mention of that? For standardization these combinations will be verified, but as extended conformance?

Reply:
>  but as extended conformance so not a MUST?

OK sounds good

Reply:
Circling back - I opt'd to leave the `MUST` requirement that implementations look at the `appProtocol` field to infer backend protocol.

And I changed the specific `support {x} protocol with {y} route` an optional (MAY) requirement